### DISCUSSION TITLE
Feature request: Support for Indicator constraints and SOS1 constraints

### DISCUSSION THREAD

**Discussion by jhelgert** [ERGO-Code/HiGHS#858] [Ideas]
Hi,

one of my favourite features of solvers like Gurobi, Cplex and SCIP is the support for indicator constraints, i.e. constraints of the form

``` none
If z = f, then a^T * x <= b    (1)
If z = f, then a^T * x  = b    (2)
If z = f, then a^T * x >= b    (3) 

where f ∈ {0,1}
```

It's a convenient way to model if-then decisions in MIPs without resorting to big-M formulations. Gurobi, Cplex and SCIP offer methods to add special constraints (SOS1, Indicator, PWL, ..) and also extend the MPS file format in several ways including Indicator constraints, see [here](https://www.ibm.com/docs/en/icos/12.10.0?topic=standard-special-records-in-mps-files-cplex-extensions). 

Are there any plans to add support for such constraints?

PS: SCIP seems to handle indicator constraints internally by the following reformulation (1), see [here](https://www.scipopt.org/doc-7.0.0/html/cons__indicator_8h.php):

``` none
a^T * x - s <= b,  
          s >= 0, 
      s + z <= 1  ( SOS1(s, z) )
```

**Comment by lgottwald** [comment]
Hey, yes we plan to add support for indicators and SOS constraints eventually, but not for the first release.

Since indicators can be formulated with big-M formulation that usually perform well we don't see it as an essential feature for the first release.

For SOS constraints we don't see them as useful as long as their special structure is not exploited in the solver which requires a bit more work than just parsing the MPS extensions and is not in the scope of the first release.

**Comment by joaquimg** [comment]
An extra motivation...
I want to use HiGHS as the default and examples solver in https://github.com/joaquimg/BilevelJuMP.jl (a package for modeling and solving bilevel problems).
However, I use a lot of SOS1, because it's easier for beginners than having them chose big-M's.
My experience is that properly choosing big-M's is usually better for solving, but worse in demonstrations and initial prototyping.
Cbc's SOS1 is way too unreliable, SCIP has weird licensing so having HiGHS would be great!

**Comment by jacobsvante** [comment]
@jajhall @galabovaa Three years have passed since this was asked. What are your thoughts on adding SOS constraints now?

**Reply by jajhall** [reply]
It's still well down our list of priorities, partly because no-one is offering us funding to add them

**Reply by jacobsvante** [reply]
Gotcha. Thanks for the quick response.

**Comment by jajhall** [comment]
> SCIP has weird licensing

This has changed in the past couple of years

**Reply by jacobsvante** [reply]
Yup, since December 2022.
